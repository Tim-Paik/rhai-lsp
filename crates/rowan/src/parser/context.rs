//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
