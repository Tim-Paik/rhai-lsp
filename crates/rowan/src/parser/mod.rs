use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context)
    }
}

/// The result of parsing.
#[derive(Debug, Clone)]
pub struct Parse {
    pub errors: Vec<ParseError>,
    pub green: GreenNode,
}

impl Parse {
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

#[derive(Debug, Clone, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    pub range: rowan::TextRange,
    pub kind: ParseErrorKind,
}

impl ParseError {
    pub fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

#[derive(Debug, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .intersperse(String::from(", "))
            .collect::<String>()
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
