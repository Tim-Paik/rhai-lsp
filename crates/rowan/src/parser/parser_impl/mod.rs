#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
