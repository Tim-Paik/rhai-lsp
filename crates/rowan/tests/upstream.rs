use rhai_rowan::parser::Parser;
use test_case::test_case;

#[test_case(include_str!("../../../rhai/scripts/array.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/assignment.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/comments.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/fibonacci.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl4.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/loop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/mat_mul.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/module.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/oop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/primes.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/speed_test.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/string.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/strings_map.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/switch.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/while.rhai"))]
fn parse_rhai_upstream(src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty());
    // TODO(tests)
    // insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
}
