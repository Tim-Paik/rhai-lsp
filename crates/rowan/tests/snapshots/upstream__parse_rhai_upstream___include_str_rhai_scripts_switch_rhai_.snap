---
source: crates/rowan/tests/upstream.rs
expression: "format!(\"{:#?}\", parse.into_syntax())"

---
FILE@0..324
  STMT@0..52
    ITEM@0..51
      EXPR@0..51
        EXPR_LET@0..51
          KW_LET@0..3 "let"
          WHITESPACE@3..4 " "
          IDENT@4..7 "arr"
          WHITESPACE@7..8 " "
          OP_ASSIGN@8..9 "="
          EXPR@9..51
            WHITESPACE@9..10 " "
            EXPR_ARRAY@10..51
              PUNCT_BRACKET_START@10..11 "["
              EXPR@11..13
                EXPR_LIT@11..13
                  LIT@11..13
                    LIT_INT@11..13 "42"
              PUNCT_COMMA@13..14 ","
              WHITESPACE@14..15 " "
              EXPR@15..22
                EXPR_LIT@15..22
                  LIT@15..22
                    LIT_FLOAT@15..22 "123.456"
              PUNCT_COMMA@22..23 ","
              WHITESPACE@23..24 " "
              EXPR@24..31
                EXPR_LIT@24..31
                  LIT@24..31
                    LIT_STR@24..31 "\"hello\""
              PUNCT_COMMA@31..32 ","
              WHITESPACE@32..33 " "
              EXPR@33..37
                EXPR_LIT@33..37
                  LIT@33..37
                    LIT_BOOL@33..37 "true"
              PUNCT_COMMA@37..38 ","
              WHITESPACE@38..39 " "
              EXPR@39..42
                EXPR_LIT@39..42
                  LIT@39..42
                    LIT_CHAR@39..42 "'x'"
              PUNCT_COMMA@42..43 ","
              WHITESPACE@43..44 " "
              EXPR@44..47
                EXPR_LIT@44..47
                  LIT@44..47
                    LIT_INT@44..47 "999"
              PUNCT_COMMA@47..48 ","
              WHITESPACE@48..49 " "
              EXPR@49..50
                EXPR_LIT@49..50
                  LIT@49..50
                    LIT_INT@49..50 "1"
              PUNCT_BRACKET_END@50..51 "]"
    PUNCT_SEMI@51..52 ";"
  WHITESPACE@52..54 "\n\n"
  STMT@54..324
    ITEM@54..323
      EXPR@54..323
        EXPR_FOR@54..323
          KW_FOR@54..57 "for"
          PAT@57..62
            WHITESPACE@57..58 " "
            PAT_IDENT@58..62
              IDENT@58..62 "item"
          WHITESPACE@62..63 " "
          KW_IN@63..65 "in"
          EXPR@65..70
            WHITESPACE@65..66 " "
            EXPR_IDENT@66..70
              IDENT@66..69 "arr"
              WHITESPACE@69..70 " "
          EXPR_BLOCK@70..323
            PUNCT_BRACE_START@70..71 "{"
            WHITESPACE@71..76 "\n    "
            STMT@76..322
              ITEM@76..322
                EXPR@76..322
                  EXPR_SWITCH@76..321
                    KW_SWITCH@76..82 "switch"
                    EXPR@82..88
                      WHITESPACE@82..83 " "
                      EXPR_IDENT@83..88
                        IDENT@83..87 "item"
                        WHITESPACE@87..88 " "
                    SWITCH_ARM_LIST@88..321
                      PUNCT_BRACE_START@88..89 "{"
                      WHITESPACE@89..98 "\n        "
                      SWITCH_ARM@98..124
                        EXPR@98..101
                          EXPR_LIT@98..100
                            LIT@98..100
                              LIT_INT@98..100 "42"
                          WHITESPACE@100..101 " "
                        PUNCT_ARROW_FAT@101..103 "=>"
                        EXPR@103..124
                          EXPR_CALL@103..124
                            EXPR@103..109
                              WHITESPACE@103..104 " "
                              EXPR_IDENT@104..109
                                IDENT@104..109 "print"
                            ARG_LIST@109..124
                              PUNCT_PAREN_START@109..110 "("
                              EXPR@110..123
                                EXPR_LIT@110..123
                                  LIT@110..123
                                    LIT_STR@110..123 "\"The Answer!\""
                              PUNCT_PAREN_END@123..124 ")"
                      PUNCT_COMMA@124..125 ","
                      WHITESPACE@125..134 "\n        "
                      SWITCH_ARM@134..179
                        EXPR@134..142
                          EXPR_LIT@134..141
                            LIT@134..141
                              LIT_FLOAT@134..141 "123.456"
                          WHITESPACE@141..142 " "
                        PUNCT_ARROW_FAT@142..144 "=>"
                        EXPR@144..179
                          EXPR_CALL@144..179
                            EXPR@144..150
                              WHITESPACE@144..145 " "
                              EXPR_IDENT@145..150
                                IDENT@145..150 "print"
                            ARG_LIST@150..179
                              PUNCT_PAREN_START@150..151 "("
                              EXPR@151..178
                                EXPR_LIT@151..178
                                  LIT@151..178
                                    LIT_STR@151..178 "`Floating point... ${ ..."
                              PUNCT_PAREN_END@178..179 ")"
                      PUNCT_COMMA@179..180 ","
                      WHITESPACE@180..189 "\n        "
                      SWITCH_ARM@189..223
                        EXPR@189..197
                          EXPR_LIT@189..196
                            LIT@189..196
                              LIT_STR@189..196 "\"hello\""
                          WHITESPACE@196..197 " "
                        PUNCT_ARROW_FAT@197..199 "=>"
                        EXPR@199..223
                          EXPR_CALL@199..223
                            EXPR@199..205
                              WHITESPACE@199..200 " "
                              EXPR_IDENT@200..205
                                IDENT@200..205 "print"
                            ARG_LIST@205..223
                              PUNCT_PAREN_START@205..206 "("
                              EXPR@206..222
                                EXPR_LIT@206..222
                                  LIT@206..222
                                    LIT_STR@206..222 "`${item} world!`"
                              PUNCT_PAREN_END@222..223 ")"
                      PUNCT_COMMA@223..224 ","
                      WHITESPACE@224..233 "\n        "
                      SWITCH_ARM@233..266
                        EXPR@233..237
                          EXPR_LIT@233..236
                            LIT@233..236
                              LIT_INT@233..236 "999"
                          WHITESPACE@236..237 " "
                        PUNCT_ARROW_FAT@237..239 "=>"
                        EXPR@239..266
                          EXPR_CALL@239..266
                            EXPR@239..245
                              WHITESPACE@239..240 " "
                              EXPR_IDENT@240..245
                                IDENT@240..245 "print"
                            ARG_LIST@245..266
                              PUNCT_PAREN_START@245..246 "("
                              EXPR@246..265
                                EXPR_LIT@246..265
                                  LIT@246..265
                                    LIT_STR@246..265 "`A number: ${item}`"
                              PUNCT_PAREN_END@265..266 ")"
                      PUNCT_COMMA@266..267 ","
                      WHITESPACE@267..276 "\n        "
                      SWITCH_ARM@276..320
                        EXPR@276..278
                          EXPR_LIT@276..277
                            LIT@276..277
                              LIT_INT@276..277 "_"
                          WHITESPACE@277..278 " "
                        PUNCT_ARROW_FAT@278..280 "=>"
                        EXPR@280..320
                          EXPR_CALL@280..315
                            EXPR@280..286
                              WHITESPACE@280..281 " "
                              EXPR_IDENT@281..286
                                IDENT@281..286 "print"
                            ARG_LIST@286..315
                              PUNCT_PAREN_START@286..287 "("
                              EXPR@287..314
                                EXPR_LIT@287..314
                                  LIT@287..314
                                    LIT_STR@287..314 "`Something else: <${i ..."
                              PUNCT_PAREN_END@314..315 ")"
                          WHITESPACE@315..320 "\n    "
                      PUNCT_BRACE_END@320..321 "}"
                  WHITESPACE@321..322 "\n"
            PUNCT_BRACE_END@322..323 "}"
    WHITESPACE@323..324 "\n"

